[FLINK-33138] DataStream API implementation

nicusX opened the pull request:
## Purpose of the change

DataStream API implementation

Also includes:
* AMP Request Signed implementation ([FLINK-33141](https://issues.apache.org/jira/browse/FLINK-33141))
* Sample application, running on Amazon Managed Service for Apache Flink and writing to Amazon Managed Prometheus 

## Verifying this change

* Unit tests
* Run included sample application on Amazon Managed Service for Apache Flink

### Significant changes

(Please check any boxes [x] if the answer is "yes" )

 * [ ] Dependencies have been added or upgraded
 * [ ] Public API has been changed (Public API is any class annotated with @Public(Evolving))
 * [ ] Serializers have been changed
 * [x] New feature has been introduce
     * If yes, how is this documented? JavaDocs and README.md

nicusX commented:
@hlteoh37 @dannycranmer please review this PR when you have the chance, so we can proceed with the other related tasks

hlteoh37 commented on amp-request-signer/pom.xml:
```diff
@@ -0,0 +1,63 @@
+        <maven.compiler.source>${target.java.version}</maven.compiler.source>
+        <maven.compiler.target>${target.java.version}</maven.compiler.target>
+
+        <flink.version>1.16.0</flink.version>
```
Should we consider 1.17? Since the supported version is 1.17 + 1.18 now

hlteoh37 commented on amp-request-signer/pom.xml:
```diff
@@ -0,0 +1,63 @@
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.flink</groupId>
+            <artifactId>flink-connector-prometheus</artifactId>
+            <version>${project.version}</version>
+        </dependency>
```
IIUC, this should be `<scope>provided</scope>`!

hlteoh37 commented on amp-request-signer/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AWS4SignerBase.java:
```diff
@@ -0,0 +1,290 @@
+
+import com.amazonaws.util.BinaryUtils;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
```
Is there a reason we use `javax.crypto.Mac` here? Do we know if this limits our java version etc?

nicusX replied:
javax.crypto.Mac was introduced in Java 1.4 and is still included in Java 18. I don't think we are limiting any versions.

nicusX replied:
The whole signer implementation is a piece of code that half a dozen OSS projects are using, practically copy&paste, because nobody published it as a library
Here is just one example: https://github.com/apache/jclouds/blob/master/apis/s3/src/main/java/org/jclouds/s3/filters/Aws4SignerBase.java

hlteoh37 commented on amp-request-signer/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AWS4SignerBase.java:
```diff
@@ -0,0 +1,290 @@
+        this.serviceName = serviceName;
+        this.regionName = regionName;
+
+        dateTimeFormat = new SimpleDateFormat(ISO_8601_BASIC_FORMAT);
+        dateTimeFormat.setTimeZone(new SimpleTimeZone(0, "UTC"));
+        dateStampFormat = new SimpleDateFormat(DATE_STRING_FORMAT);
+        dateStampFormat.setTimeZone(new SimpleTimeZone(0, "UTC"));
```
Is there a reason we fix this as 0 UTC? Is there a scenario we would want to have a different timezone here? If so, maybe we can make this configurable / use the timezone of the host

nicusX replied:
This formatter is used to build the `x-amz-date` request header that must be in UTC, as documented here https://docs.aws.amazon.com/IAM/latest/UserGuide/signing-elements.html#date

hlteoh37 commented on amp-request-signer/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AmazonManagedPrometheusWriteRequestSigner.java:
```diff
@@ -0,0 +1,94 @@
+import java.net.URL;
+import java.util.Map;
+
+/** Sing a Remote-Write request to Amazon Managed Service for Prometheus (AMP). */
```
nit: `Sign`

We also might want to put an example usage in the docs here.

nicusX replied:
Added an example of usage in the signer README

hlteoh37 commented on amp-request-signer/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AmazonManagedPrometheusWriteRequestSigner.java:
```diff
@@ -0,0 +1,94 @@
+     * @param awsRegion Region of the AMP workspace
+     */
+    public AmazonManagedPrometheusWriteRequestSigner(String remoteWriteUrl, String awsRegion) {
+        Preconditions.checkArgument(StringUtils.isNotBlank(awsRegion));
```
Should we add a message here in the event of failure?

nicusX replied:
Added message for both parameters.
Also using `isNotBlank()` with both, for consistency

hlteoh37 commented on msf-amp-example/README.md:
```diff
@@ -0,0 +1,68 @@
+## Sample application: Amazon Managed Service for Apache Flink and Amazon Managed Prometheus
```
We shouldn't be putting company specific examples here. Should we consider using a non-MSF example here instead, and move this MSF-specific one elsewhere?

nicusX replied:
Replaced the example with a new one, using Flink 1.17.2 and with nothing MSF specific.
The new example optionally demonstrates how to configure the request signer to target AMP (the only supported request signer at the moment)

nicusX replied:
Updated to 1.18.1

hlteoh37 commented on pom.xml:
```diff
@@ -0,0 +1,82 @@
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.apache.flink</groupId>
+    <artifactId>flink-prometheus</artifactId>
```
Should we keep the same convention as other connector repos `flink-connector-prometheus-parent`?

nicusX replied:
Updated

hlteoh37 commented on prometheus-connector/pom.xml:
```diff
@@ -0,0 +1,130 @@
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.8.0</version>
+                <configuration>
+                    <source>${target.java.version}</source>
+                    <target>${target.java.version}</target>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
```
this might already come from the connector parent! We should check!

nicusX replied:
Moved to parent only

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusRemoteWriteHttpRequestBuilder.java:
```diff
@@ -0,0 +1,70 @@
+                    "X-Prometheus-Remote-Write-Version",
+                    "0.1.0",
+                    HttpHeaders.USER_AGENT,
+                    "Flink-Prometheus/0.1.0" // TODO Prometheus requires a user-agent header. What
```
As discussed offline, maybe we can make this configurable, and default to `Flink-Prometheus`?

nicusX replied:
Made user-agent configurable with fallback default

hlteoh37 reviewed (commented):
Initial comments

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkWriter.java:
```diff
@@ -0,0 +1,267 @@
+                new ResponseCallback(timeSeriesCount, sampleCount, counters, requestResult));
+    }
+
+    @VisibleForTesting
```
Not a fan of this `@VisibleForTesting` - can we instead write a test  `asyncHttpClient` that tests each scenario?

nicusX replied:
Moved the callback to a separate class and created a specific test.

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkBuilder.java:
```diff
@@ -0,0 +1,203 @@
+    @Override
+    public PrometheusSinkBuilder setMaxBatchSizeInBytes(long maxBatchSizeInBytes) {
+        throw new UnsupportedOperationException(
+                "maxBatchSizeInBytes is not supported by this sink");
```
Should we add a comment on the reason why we disable it here?

hlteoh37 replied:
After offline discussion, maybe we can do something like 

```
Use setMaxBatchSizeInSamples() instead
```

nicusX replied:
Added comments to all disabled parameters, explaining what to use instead

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkWriter.java:
```diff
@@ -0,0 +1,267 @@
+                            timeSeriesCount,
+                            sampleCount);
+                    counters.inc(NUM_SAMPLES_RETRY_LIMIT_DROPPED, sampleCount);
+                }
```
Should we add a default log here? In the event that `RemoteWriteResponseClassifier.isRetriableErrorResponse` and `RemoteWriteResponseClassifier.isNonRetriableErrorResponse` don't cover the entire set of responses?

nicusX replied:
Added a different behaviour: no response should be unclassified. If a status code < 200 is received, something is terribly wrong, so I am now throwing an exception

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkWriter.java:
```diff
@@ -0,0 +1,267 @@
+            requestResult.accept(Collections.emptyList());
+        }
+
+        @Override
+        public void failed(Exception ex) {
+            LOG.warn(
+                    "Exception executing the remote-write (discarded {} time-series containing {} samples)",
+                    timeSeriesCount,
+                    sampleCount,
+                    ex);
+            counters.inc(NUM_SAMPLES_DROPPED, sampleCount);
+            counters.inc(NUM_WRITE_REQUESTS_PERMANENTLY_FAILED);
+        }
```
Should we retry these failures / Propagate the exception up to transition the job to failed?

nicusX replied:
Changed the general behaviour (see comment below)

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkWriter.java:
```diff
@@ -0,0 +1,267 @@
+        }
+
+        @Override
+        public void cancelled() {
+            LOG.warn(
+                    "Write request execution cancelled (discarded {} time-series containing {} samples)",
+                    timeSeriesCount,
+                    sampleCount);
+            counters.inc(NUM_SAMPLES_DROPPED, sampleCount);
+            counters.inc(NUM_WRITE_REQUESTS_PERMANENTLY_FAILED);
+        }
```
I wonder if we should also propagate the exception upwards here rather than swallowing it. For example, this might be triggered on interrupt / non-graceful shutdown of the async Http client

nicusX replied:
Done: always throws an exception on cancelled

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkWriter.java:
```diff
@@ -0,0 +1,267 @@
+                            sampleCount);
+                    counters.inc(NUM_SAMPLES_NON_RETRIABLE_DROPPED, sampleCount);
+                } else if (RemoteWriteResponseClassifier.isRetriableErrorResponse(response)) {
+                    LOG.warn(
```
We shouldn't be dropping records here on a retryable error.

IMO we should either requeue the records, or propagate up the exception to fail the job.

hlteoh37 replied:
After offline discussion, we agreed that maybe we can set the HttpCllient to a lower maxRetryCount (by default), and then if the maxRetry is exceeded, we can throw the Exception to restart the Flink job

nicusX replied:
Modified the behaviour so that, by default, it will throw an exception and stop the job on any unrecoverable condition. Optionally, the user can decide not to block the job, but drop data and continue, selectively for three separate conditions:
1. Max retry exceeded
2. Non-retriable error reported
3. Http client I/O error
When the Http request is cancelled, an exception is always thrown (not configurable)

The default max retry is now 100, ~ 500 sec with the default backoff timing

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusTimeSeries.java:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    /** A single Sample. */
+    public static class Sample implements Serializable {
```
we should override equals and hashcode here too

nicusX replied:
implemented

hlteoh37 commented on prometheus-connector/src/test/java/org/apache/flink/connector/prometheus/sink/PrometheusRemoteWriteHttpRequestBuilderTest.java:
```diff
@@ -0,0 +1,76 @@
+
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
```
We don't recommend using Mockito in Flink. Usually we prefer creating test constructs to test actual behavior,

https://flink.apache.org/how-to-contribute/code-style-and-quality-common/#avoid-mockito---use-reusable-test-implementations

nicusX replied:
Removed Mockito from all tests.

nicusX replied:
There is still a test with WireMock to test the http request handling stack, including the async http client behaviour on specific responses, that would have been otherwise untestable.

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSink.java:
```diff
@@ -0,0 +1,117 @@
+import java.util.Collection;
+
+/** Sink implementation accepting {@link PrometheusTimeSeries} as inputs. */
+public class PrometheusSink extends AsyncSinkBase<PrometheusTimeSeries, Types.TimeSeries> {
```
Let's use Flink's `@Public` annotation here

nicusX replied:
Marked as @Public all the classes representing the public interface of the connector: `PrometheusSink`, `PrometheusTimeSeries`, and `PrometheusRequestSigner`

hlteoh37 replied:
Sorry!! I think better to use `@PublicEvolving` or `@Experimental`, either ok with me. Prefer `@Experimental` because it allows easier breaking changes on our side.

hlteoh37 reviewed (changes requested):
Added some comments after 2nd parse!

nicusX commented:
@hlteoh37 please check out the changes. I should have addressed all the points

hlteoh37 commented on amp-request-signer/pom.xml:
```diff
@@ -0,0 +1,60 @@
+        <dependency>
+            <groupId>com.amazonaws</groupId>
+            <artifactId>aws-java-sdk-core</artifactId>
+            <version>1.12.570</version>
```
can we delegate this to a bom in parent?

nicusX replied:
Added SDKv1 BOM to the parent

hlteoh37 commented on example-datastream-job/pom.xml:
```diff
@@ -0,0 +1,156 @@
+        <!-- For Flink local development -->
+        <dependency>
+            <groupId>org.apache.flink</groupId>
+            <artifactId>flink-runtime-web</artifactId>
```
why do we need flink-runtime-web?

nicusX replied:
Removed the Flink dashboard when running locally. Removed this dependency, that was used for that

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/errorhandling/SinkWriterErrorHandlingBehaviorConfiguration.java:
```diff
@@ -0,0 +1,105 @@
+            return this;
+        }
+
+        public Builder onPrometheusNonRetriableError(OnErrorBehavior onErrorBehavior) {
+            this.onPrometheusNonRetriableError = onErrorBehavior;
+            return this;
```
we should consider making this non-configurable

nicusX replied:
Absolutely not. For Prometheus, non-retriable errors are malformed or out-of-order writes. This may happen because of input data, and the writer may want to just log, drop, and continue. This is actually the default behaviour of most of tools writing to Prometheus.

hlteoh37 commented on example-datastream-job/README.md:
```diff
@@ -0,0 +1,22 @@
+
+It generates random dummy Memory and CPU metrics from a number of instances, and writes them to Prometheus.
+
+### Configuration
+
+The application expects these parameters, via command line:
+
+* `--prometheusRemoteWriteUrl <URL>`: the Prometheus remote-write URL to target
+* `--awsRegion <region>`: (optional) if specified, it configures the Amazon Managed Prometheus request signer for a workspace in this Region
+* `--webUI`: (optional, for local development only) enables Flink Web UI, with flame graphs, for local development
```
How do we run this? Should we include an example run command?

nicusX replied:
I don't think that would be any useful.
The way you pass parameters depends on your deployment.
And if you are running locally in your IDE, it depends on your IDE.

hlteoh37 commented on example-datastream-job/README.md:
```diff
@@ -0,0 +1,22 @@
+
+The application generates random time series, containing `CPU` and `Memory` samples from 5 dummy "instances". A new time series is generated about every 100ms. Each time series contains 1 to 10 samples.
+
+These parameters are configurable from the code
```
```suggestion
These parameters are configurable from the code.
```

hlteoh37 commented on example-datastream-job/src/main/java/org/apache/flink/connector/prometheus/examples/DataStreamJob.java:
```diff
@@ -0,0 +1,140 @@
+            env = StreamExecutionEnvironment.getExecutionEnvironment();
+        }
+
+        env.setParallelism(2);
```
Can we explain why this is set to 2?

nicusX replied:
Added a comment.
The goal is to force parallel writes and demonstrate how `PrometheusTimeSeriesLabelsAndMetricNameKeySelector` partitions data preventing out-of-order writes, that would be rejected by Prometheus

hlteoh37 commented on example-datastream-job/src/main/java/org/apache/flink/connector/prometheus/examples/DataStreamJob.java:
```diff
@@ -0,0 +1,140 @@
+        // Configure data generator
+        int generatorMinSamplesPerTimeSeries = 1;
+        int generatorMaxSamplesPerTimeSeries = 10;
+        int generatorNumberOfDummyInstances = 5;
```
Hmm, this is hard to understand out of context. Can we add an overall description at the top of this class, or rename this to "NumberOfKeys/NumberOfUniqueMetrics" or something?

nicusX replied:
Reworked the random data generator to generate a configurable random number of metrics (`M0000`, `M00001`...) for a configurable random number of sources (`SourceID` = `S0000000000`, `S0000000001`...).
Removed confusing references to "instances", "CPU" and "Memory"

hlteoh37 commented on example-datastream-job/src/main/java/org/apache/flink/connector/prometheus/examples/DataStreamJob.java:
```diff
@@ -0,0 +1,140 @@
+                        .generator();
+
+        SourceFunction<PrometheusTimeSeries> source =
+                new FixedDelayDataGenertorSource<>(
```
Typo - `Generator`

nicusX replied:
Fixed

hlteoh37 commented on example-datastream-job/src/main/java/org/apache/flink/connector/prometheus/examples/DataStreamJob.java:
```diff
@@ -0,0 +1,140 @@
+        // value
+        AsyncSinkBase<PrometheusTimeSeries, Types.TimeSeries> sink =
+                PrometheusSink.builder()
+                        .setMaxBatchSizeInSamples(500)
+                        .setMaxRecordSizeInSamples(500)
+                        .setMaxTimeInBufferMS(5000)
+                        .setRetryConfiguration(
+                                RetryConfiguration.builder()
+                                        .setInitialRetryDelayMS(30L)
+                                        .setMaxRetryDelayMS(5000L)
+                                        .setMaxRetryCount(100)
+                                        .build())
+                        .setSocketTimeoutMs(5000)
```
nit: Should we leave comment after to show which are optional?
e.g.
```
   .setMaxBatchSizeInSamples(500) // Optional, defaults to 500
```

hlteoh37 commented on example-datastream-job/src/main/java/org/apache/flink/connector/prometheus/examples/SimpleInstanceMetricsTimeSeriesGenerator.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    private String dummyInstanceId(int number) {
+        return "I" + String.format("%010d", number);
```
nit: instance ID is lower case `i`

hlteoh37 commented on example-datastream-job/src/main/resources/log4j2.properties:
```diff
@@ -0,0 +1,29 @@
+appender.console.layout.type = PatternLayout
+appender.console.layout.pattern = %d{HH:mm:ss,SSS} %-5p %-60c %x - %m%n
+
+#logger.tracing.name = org.apache.flink.connector.prometheus.sink
+#logger.tracing.level = trace
+#logger.tracing.additivity = false
+#logger.tracing.appenderRef.console.ref = ConsoleAppender
```
Can we remove these, or explain when to uncomment for testing?

hlteoh37 commented on pom.xml:
```diff
@@ -0,0 +1,120 @@
+            <dependency>
+                <groupId>com.amazonaws</groupId>
+                <artifactId>aws-java-sdk-bom</artifactId>
+                <version>${aws.sdkv1.version}</version>
```
Why are we using AWS SDK v1? Can we use v2?

nicusX replied:
This is actually no longer needed. The AMP request signer no longer uses any SDK
I am removing it

nicusX replied:
Correction: it is still required. But I am moving any reference to AWS SDK dependency to the AMP Signer module.

This is used to sign the AMP request only. There is no working example using SDK2 in Java. We managed to make it work with SDKv1. I'd suggest to keep it as-is for the moment and consider upgrading the signer in a future version.

hlteoh37 commented on prometheus-connector/README.md:
```diff
@@ -0,0 +1,189 @@
+of Prometheus Remote-Write, the sink guarantees that input data are written to Prometheus only if input data are in order and well-formed.
+
+For efficiency, the connector does not do any validation.
+If input is out of order or malformed, the write request is rejected by Prometheus and data is discarded by the sink.
```
We can mention our sink configurations

nicusX replied:
Configuration is extensively documented below, in the README

nicusX replied:
Sorry, understood the point. Adding an explanation

hlteoh37 commented on prometheus-connector/README.md:
```diff
@@ -0,0 +1,189 @@
+On non-retriable error response (4xx, except 429, non retryable exceptions), or on reaching the retry limit, 
+**the entire write-request**, containing the batch of time-series, **is dropped**.
+
+Every dropped request is logged at WARN level, including the reason provided by the remote-write endpoint.
```
We can mention the configuration here too

nicusX replied:
Added explanation

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/HttpResponseCallback.java:
```diff
@@ -0,0 +1,192 @@
+    private final SinkWriterErrorHandlingBehaviorConfiguration errorHandlingBehaviorConfig;
+
+    public HttpResponseCallback(
+            int timeSeriesCount,
+            long sampleCount,
+            SinkMetrics metrics,
```
nit: This can be abstracted away to another `metricsCallback`

nicusX replied:
Not exactly a minor change, because it has impact in many places. But definitely a good idea.
Implemented.

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/HttpResponseCallback.java:
```diff
@@ -0,0 +1,192 @@
+     * handling configuration.
+     */
+    @Override
+    public void completed(SimpleHttpResponse response) {
```
Do we want to consider expanding out nested `if-else` into linear with `return` instead?

See example here: https://stackoverflow.com/questions/56356190/nested-if-conditions-vs-multiple-separated-if-conditions-with-return-statement-i

nicusX replied:
Done. 
One nested if (no else) is still required to throw exception when the error behaviour is FAIL, avoiding too many `if (A && X1)... if (A && X2)...`

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusRequestSigner.java:
```diff
@@ -0,0 +1,43 @@
+     *     expected to modify the Map in place. The Map is expected to be mutable.
+     * @param requestBody request body, already compressed.
+     */
+    void addSignatureHeaders(Map<String, String> requestHeaders, byte[] requestBody);
```
Let's add an explanation here as to why we mutate request headers instead of creating a copy - for performance reasons

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSink.java:
```diff
@@ -0,0 +1,129 @@
+    private final String httpUserAgent;
+    private final SinkWriterErrorHandlingBehaviorConfiguration errorHandlingBehaviorConfig;
+
+    protected PrometheusSink(
```
Can we add a configuration to specify a metrics group prefix?

nicusX replied:
Added a configurable metric group that gets appended after the `Sink__Writer` of AsyncSinkBase. I can't see how I can add a prefix

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkBuilder.java:
```diff
@@ -0,0 +1,246 @@
+    @Override
+    public AsyncSinkBase<PrometheusTimeSeries, Types.TimeSeries> build() {
+
+        int maxInFlightRequest = 1;
```
nit: this could be a static constant. (I understand this is a copy of the real value in the sink itself)

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusStateSerializer.java:
```diff
@@ -0,0 +1,154 @@
+                long requestEntrySize =
+                        RequestEntrySizeUtils.requestSerializedSize(wrapper.getRequestEntry());
+                out.writeLong(requestEntrySize); // SIZEn
+                serializeRequestToStream(wrapper.getRequestEntry(), out); // REQUESTn
```
remove the `n`?

hlteoh37 commented on amp-request-signer/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AWS4SignerForAuthorizationHeader.java:
```diff
@@ -0,0 +1,124 @@
+ * Sample AWS4 signer demonstrating how to sign requests to Amazon S3 using an 'Authorization'
+ * header.
+ */
+public class AWS4SignerForAuthorizationHeader extends AWS4SignerBase {
```
Can we add unit tests?

nicusX replied:
It's silly, but this "just works". There is no specs . 
This is a class copied&pasted in a gazillion of OSS projects, including big ones.
(try searching "Sample AWS4 signer demonstrating how to sign requests to Amazon S3" in GitHub)
Any unit test would be written reverse-engineering the implementation, so useless.

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/prometheus/Remote.java:
```diff
@@ -0,0 +1,6661 @@
+/*
```
Should we consider adding API snapshot tests for this class?

hlteoh37 replied:
Same for `Types.java` and `GoGoProtos.java`

z3d1k replied:
Should original `*.proto` files also be added here? With a way to generate java files.
This way it will be clearer how to update this code if API change in the future releases.

nicusX replied:
I think we discussed this.
Generating the Java classes from proto cannot be done with a pure Java solution included in the project. 
It requires ProtoBuf binaries. (e.g. see https://github.com/javiroman/jremotewrite/blob/e1b5d9865d42254b32f2da6ddfd6956569577601/pom.xml#L210)

 I am adding the *.proto file for reference and a section to the README to explain

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusTimeSeries.java:
```diff
@@ -0,0 +1,186 @@
+            return value;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            Label label = (Label) o;
+            return new EqualsBuilder()
+                    .append(name, label.name)
+                    .append(value, label.value)
+                    .isEquals();
+        }
```
Can we add unit test coverage for this?

nicusX replied:
Added unit tests on equals and hashCode semantics, and also fixed a bug: missing equals() in one of the inner classes!

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusTimeSeries.java:
```diff
@@ -0,0 +1,186 @@
+        private List<Sample> samples = new ArrayList<>();
+        private String metricName;
+
+        private Builder(List<Label> labels, List<Sample> samples, String metricName) {
+            this.labels = labels;
+            this.samples = samples;
+            this.metricName = metricName;
+        }
```
Can we add unit test coverage for this? 
<img width="717" alt="image" src="https://github.com/apache/flink-connector-prometheus/assets/35062175/e03c8c41-bf9b-4339-ab0f-34ac4b62b241">

nicusX replied:
I assume you mean testing the builder, not its constructor.
Added a test for PrometheusTimeSeries.Builder

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/RequestEntrySizeUtils.java:
```diff
@@ -0,0 +1,61 @@
+     * @param requestEntries collection of time-series
+     * @return number of samples
+     */
+    public static long countSamples(Collection<Types.TimeSeries> requestEntries) {
+        return requestEntries.stream()
+                .mapToLong(RequestEntrySizeUtils::requestSizeForBatching)
+                .sum();
+    }
```
Can we add unit test coverage for this?

nicusX replied:
Added tests for RequestEntrySizeUtilsTest methods

hlteoh37 commented on prometheus-connector/src/test/java/org/apache/flink/connector/prometheus/sink/PrometheusStateSerializerTest.java:
```diff
@@ -0,0 +1,108 @@
+    // the requestEntry "size"
+    // (i.e. the number of Samples). This is the "size" used in RequestEntryWrapper
+    // see
+    // https://github.com/apache/flink/blob/69e812688b43be9a0c4f79e6af81bc2d1d8a873e/flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterStateSerializer.java#L60
```
Can we consider making this a javadoc + link to code?
For example:
```
    /**
     * This method uses the same implementation as PrometheusSinkWriter.getSizeInBytes() to extract
     * the requestEntry "size"
     * (i.e. the number of Samples). This is the "size" used in RequestEntryWrapper
     * see {@link org.apache.flink.connector.base.sink.writer.AsyncSinkWriterStateSerializer#serialize(BufferedRequestState)}
     */
```

hlteoh37 replied:
(we can replace the qualified name with an import for the class)

nicusX replied:
Replaced with @link to the method.
But I am not convinced it makes sense: the link was to a specific source code to show how the size is used in the implementation of a method, not to the JavaDoc of that method

nicusX replied:
Reviewing this, I put back the link to the implementation in GitHub.
What we want to show is the implementation, not the javadoc.

hlteoh37 commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusTimeSeries.java:
```diff
@@ -0,0 +1,186 @@
+        private final String name;
+        private final String value;
+
+        public Label(String name, String value) {
```
Can we add some `null`/`empty` validations for these data classes? (+ unit  tests too!)

nicusX replied:
There other other validations that are more critical, like Labels must be in order by name and samples in order by timestamp, or Prometheus will reject them.
We kept these validations out of the implementation on purpose, because this object will be called a gazillion times, and this may impact performance.
In the docs, we explicitly say this is a responsibility of the application, like sending the time series in time order.

Either we decide to add validation to this class, paying the performance fee, or we leave this class unvalidated

nicusX replied:
I added further explanations in the README about the behavior and the responsibility of sending well-formed data.
In any case, with the default error-handling behavior, if the application sends malformed data, the sink throws an exception when Prometheus rejects the write.

hlteoh37 reviewed (changes requested):
Generally, the code structure looks good! Thanks @nicusX and @FranMorilloAWS

hlteoh37 commented:
Once all the above are fixed, it would be good if we can squash the commits!

hlteoh37 commented on prometheus-connector/src/test/java/org/apache/flink/connector/prometheus/sink/InspectableMetricGroupAssertions.java:
```diff
@@ -0,0 +1,63 @@
+                actualCounterCount,
+                "The counter "
+                        + counterName
+                        + " was expected tp be "
```
```suggestion
                        + " was expected to be "
```

z3d1k commented on amp-request-signer/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AWS4SignerBase.java:
```diff
@@ -0,0 +1,290 @@
+     * Returns the canonical request string to go into the signer process; this consists of several
+     * canonical sub-parts.
+     *
+     * @return
```
Missing description of returned value?

nicusX replied:
Removed `@return`, that was actually redundant with the description above

nicusX replied:
@z3d1k not sure I see the value of generating the Protobuf classes from `.proto` file instead of hardwiring the generated classes, in this context. 
The remote-write protocol is much more than the payload schema.
Any Prometheus remote-write protocol change will probably require changes to the code to adjust the behavior. 
Also, generating the code from `.proto` would require to do the same also with GoGoProto.

And finally, this an external dependency, not something under our control, and must be "copied&pasted" into the project source, as it's not available on Maven. There is no much difference between including the proto and including the generated code

z3d1k replied:
Discussed offline, my comment was about including _both_ generated version and source `*.proto` with a way to regenerate code.

z3d1k commented on amp-request-signer/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AmazonManagedPrometheusWriteRequestSigner.java:
```diff
@@ -0,0 +1,98 @@
+    private final String awsRegion;
+
+    /**
+     * Constructor.
+     *
```
Is this comment required?
> JavaDocs should not state meaningless information (just to satisfy the Checkstyle checker).

https://flink.apache.org/how-to-contribute/code-style-and-quality-common/#comments

nicusX replied:
Modified the comment only retaining the `@param` docs

z3d1k commented on amp-request-signer/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AmazonManagedPrometheusWriteRequestSigner.java:
```diff
@@ -0,0 +1,98 @@
+                contentHashString); // this header must be included before generating the
+        // Authorization header
+
+        DefaultAWSCredentialsProviderChain credsChain = new DefaultAWSCredentialsProviderChain();
```
Should this be made configurable? Some use cases may require specific credentials provider implementation.

nicusX replied:
Added the possibility to instantiate a signer with a different credentials provider.

nicusX replied:
Actually, I reverted the change. Just passing a credentials provider instance obviously doesn't work, not being serializable.

Making the credentials provider customisable would mean either:
1. Copying the implementation of `org.apache.flink.connector.aws.util.AWSGeneralUtil` from `flink-connector-aws-base` and using it to create the credential provider at runtime, based on configuration, or
2. Including `flink-connector-aws-base` as a dependency. But this would bring in a number of transitive dependencies that are not used. The only AWS SDK dependencies in this connector are in this signer, and only to get the credentials. Also, `flink-connector-aws-base` uses sdk2, while this signer uses sdk1. We discussed with @hlteoh37 that porting to sdk2 is nice to have but not a priority for the first release.

Considering `DefaultAWSCredentialsProviderChain` will already cover almost every possible authentication scenario when a Flink application will be writing to AMP, I would not consider configurable credentials provider a priority for the first release, and wait if any actual user will need it.

Happy to discuss.

nicusX commented:
@hlteoh37 @z3d1k please have a look, I pushes the changes (in a single squashed commit).
I implemented all the requested changes except those I explicitly commented not to implement.
Happy to discuss

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusRemoteWriteHttpRequestBuilder.java:
```diff
@@ -0,0 +1,76 @@
+            requestSigner.addSignatureHeaders(headers, httpRequestBody);
+        }
+
+        var builder =
```
Default target java version for connectors is 8, is java 11 features necessary?

nicusX replied:
Is it still the case for 1.18?
Anyhow, I removed java 11 features

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkBuilder.java:
```diff
@@ -0,0 +1,246 @@
+    private static final int DEFAULT_MAX_BATCH_SIZE_IN_SAMPLES = 500;
+    private static final long DEFAULT_MAX_TIME_IN_BUFFER_MS = 5000;
+    private static final int DEFAULT_MAX_BUFFERED_REQUESTS =
+            1000; // Max nr of requestEntry that will be buffered
```
nit: moving comment above statement will prevent formatter from splitting the line

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusRemoteWriteHttpRequestBuilder.java:
```diff
@@ -0,0 +1,76 @@
+        this.fixedHeaders =
+                Map.of(
+                        HttpHeaders.CONTENT_ENCODING,
+                        "snappy",
+                        "X-Prometheus-Remote-Write-Version",
+                        "0.1.0",
```
Can these values be moved to constant?

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkBuilder.java:
```diff
@@ -0,0 +1,246 @@
+                Optional.ofNullable(getMaxTimeInBufferMS()).orElse(DEFAULT_MAX_TIME_IN_BUFFER_MS);
+        int actualMaxRecordSizeInSamples =
+                Optional.ofNullable(getMaxRecordSizeInSamples())
+                        .orElse(getMaxBatchSizeInSamples()); // By default, set max record size to
+        // max batch size (in samples)
```
The comment repeats logic described in code. Is it required?

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkWriter.java:
```diff
@@ -0,0 +1,189 @@
+     * In our case is the max number of Samples in a single TimeSeries sent to the Sink. We are
+     * limiting the number of Samples in each TimeSeries to the max batch size, setting
+     * maxRecordSizeInBytes = maxBatchSizeInBytes.
+     */
```
Should be moved to class javadoc comment

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusStateSerializer.java:
```diff
@@ -0,0 +1,154 @@
+     * AsyncSinkWriterStateSerializer.serialize().
+     *
+     * <p>The state is serialized in form of
+     * [DATA_IDENTIFIER,NUM_OF_ELEMENTS,SIZE1,REQUEST1,SIZE2,REQUEST2....], where REQUESTn is the
```
Typo: `REQUESTn`, should be `REQUEST`

nicusX replied:
It actually means "REQUEST<n>"
Fixed the comment to make it clear

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusStateSerializer.java:
```diff
@@ -0,0 +1,154 @@
+     * <p>The state is serialized in form of
+     * [DATA_IDENTIFIER,NUM_OF_ELEMENTS,SIZE1,REQUEST1,SIZE2,REQUEST2....], where REQUESTn is the
+     * Protobuf-serialized representation of a {@link Types.TimeSeries TimeSeries}. In this
+     * implementation SIZEn is the size of the Protobuf serialization, in bytes, that does not match
```
Typo: `SIZEn`, should be `SIZE`

nicusX replied:
As above, it means "SIZE<n>".
Clarified the comment

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusStateSerializer.java:
```diff
@@ -0,0 +1,154 @@
+            List<RequestEntryWrapper<Types.TimeSeries>> serializedState = new ArrayList<>();
+            for (int i = 0; i < numberOfElements; i++) {
+                // This is the size of an request-entry in the serialized state
+                long requestSerializedSize = in.readLong(); // SIZEn - the serialized size
```
Typo: `SIZEn`, should be `SIZE`

nicusX replied:
as above

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkWriter.java:
```diff
@@ -0,0 +1,189 @@
+import java.util.function.Consumer;
+
+/** Writer, taking care of batching the {@link PrometheusTimeSeries} and handling retries. */
+public class PrometheusSinkWriter extends AsyncSinkWriter<PrometheusTimeSeries, Types.TimeSeries> {
```
Please add tests for this class, currently not covered

nicusX replied:
Not much to test here, except the initialization that is coupled with the PrometheusSink initialization.
I created a test similar to what is done in the Kinesis Sink.

https://github.com/apache/flink-connector-aws/blob/main/flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/sink/KinesisStreamsSinkWriterTest.java

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/RequestEntrySizeUtils.java:
```diff
@@ -0,0 +1,61 @@
+     * @return number of samples
+     */
+    public static long countSamples(Collection<Types.TimeSeries> requestEntries) {
+        return requestEntries.stream()
+                .mapToLong(RequestEntrySizeUtils::requestSizeForBatching)
+                .sum();
```
Should this be replaced with a loop?
https://flink.apache.org/how-to-contribute/code-style-and-quality-java/#java-streams

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/SinkMetrics.java:
```diff
@@ -0,0 +1,84 @@
+
+    /** Enum defining all sink counters. */
+    public enum SinkCounter {
+        // Total number of Samples that were dropped because of not being retriable errors in
+        // Prometheus
```
Here and below: replace comment with JavaDoc

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/http/PrometheusAsyncHttpClientBuilder.java:
```diff
@@ -0,0 +1,92 @@
+
+        final IOReactorConfig ioReactorConfig =
+                IOReactorConfig.custom()
+                        .setIoThreadCount(1) // Always one thread
```
Can you expand the comment with why a single thread is used here?

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/http/PrometheusAsyncHttpClientBuilder.java:
```diff
@@ -0,0 +1,92 @@
+        var client =
+                HttpAsyncClients.custom()
+                        .setConnectionManager(
+                                PoolingAsyncClientConnectionManagerBuilder.create()
+                                        .setDefaultTlsConfig(
+                                                TlsConfig.custom()
+                                                        .setVersionPolicy(
+                                                                HttpVersionPolicy.FORCE_HTTP_1)
+                                                        .build())
+                                        .build())
```
Should this part be extracted into variable to avoid deep nesting?

z3d1k commented on prometheus-connector/src/test/java/org/apache/flink/connector/prometheus/sink/HttpResponseHandlingBehaviourIT.java:
```diff
@@ -0,0 +1,304 @@
+    void shouldNotRetryCompleteAndThrowExceptionOn304(WireMockRuntimeInfo wmRuntimeInfo)
+            throws URISyntaxException, IOException {
+        int statusCode =
+                304; // 304, Not modified is just status code for which the behaviour is not
+        // specified by Prometheus remote-write specs, and makes the http client
+        // terminate successfully
```
Move this long comment above the expression

z3d1k commented on amp-request-signer/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AWS4SignerBase.java:
```diff
@@ -0,0 +1,288 @@
+            md.update(text.getBytes("UTF-8"));
+            return md.digest();
+        } catch (Exception e) {
+            throw new RuntimeException(
+                    "Unable to compute hash while signing request: " + e.getMessage(), e);
```
Can you replace this with a more specific exception? Same comment to all other `RuntimeException` usages

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkWriter.java:
```diff
@@ -0,0 +1,189 @@
+        try {
+            requestBody = compressWriteRequest(writeRequest);
+        } catch (IOException e) {
+            throw new RuntimeException("Exception compressing the request body", e);
```
Can you replace this with a more specific exception? Same comment to all other `RuntimeException` usages

nicusX replied:
Replaced with a specific exception. It was actually the only generic `RuntimeException` thrown

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/http/RemoteWriteRetryStrategy.java:
```diff
@@ -0,0 +1,106 @@
+import java.net.UnknownHostException;
+import java.util.List;
+
+/** Retry strategy for the http client. */
```
Please add more details on the retry strategy logic here

z3d1k commented on prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/http/RemoteWriteRetryStrategy.java:
```diff
@@ -0,0 +1,106 @@
+    @Override
+    public TimeValue getRetryInterval(
+            HttpResponse httpResponse, int execCount, HttpContext httpContext) {
+        long calculatedDelay = initialRetryDelayMs << (execCount - 1);
```
Should multiplier be configurable instead of being set to 2?
Additionally, can jitter be added here?

nicusX replied:
I don't see much value in making the backoff strategy more sophisticated. 
Prometheus Remote-write specs just say it "MUST use a backoff algorithm to prevent overwhelming the server".
Tweaking the `initialRetryIntervalMs` and `maxRetryDelayMs` will be enough to make the strategy more or less aggressive.
Also, I don't see much value in adding a jitter. There is one writer thread per subtask, and and it's very unlikely they will be synchronized when writing to the endpoint.

z3d1k commented on amp-request-signer/pom.xml:
```diff
@@ -0,0 +1,58 @@
+    </parent>
+
+    <name>Flink : Connectors : Prometheus : AMP request signer</name>
+    <groupId>org.apache.flink.connector.prometheus</groupId>
```
Is this the correct groupId?

nicusX replied:
Fixed.
Now `groupId` is `org.apache.flink` and `artifactId` is `flink-connector-prometheus-amp-request-signer`

z3d1k commented on amp-request-signer/pom.xml:
```diff
@@ -0,0 +1,58 @@
+
+    <name>Flink : Connectors : Prometheus : AMP request signer</name>
+    <groupId>org.apache.flink.connector.prometheus</groupId>
+    <artifactId>amp-request-signer</artifactId>
```
artifactId does not represent connection to prometheus connector

z3d1k reviewed (changes requested):
Please split this PR into separate ones: 
- base connector implementation
- amazon request signer

nicusX commented:
I addressed all comments, except those I explained why I didn't.
Also, now upgraded to Flink 1.18 and newer AsyncSinkBase API.

I cannot split AMP signer and connector in separate PRs at this point, because of the interdependency with the sample application. 
Also, I strongly recommend to keep the sample application in the repository, because it serves both as documentation for the user and a way to test the connector (and optionally the signer) in an actual environment. The sample application is not supposed to be released as artifact, but just to stay in the repo as source.

hlteoh37 commented on README.md:
```diff
@@ -8,6 +8,14 @@ Apache Flink is an open source stream processing framework with powerful stream-
+This repository contains the following modules
+
+* [Prometheus Connector](./prometheus-connector): Flink Prometheus Connector implementation; supports optional request signer
+* [Sample application](./example-datastream-job): Sample application showing the usage of the connector with DataStream API. It also demonstrates how to configure the request signer. 
```
Can we move this to an integ test? Adding it as a separate module would make releases more difficult.

nicusX replied:
I don't think an integration test is useful for a user. And the lack of proper working examples is a common obstacle to the adoption for users new to Flink.
This module is not supposed to be released in any form. Can't we just keep it as source, and possibly reference from the documentation?

hlteoh37 replied:
We have done it in tests previously to mitigate the "clutter" problem, and difficulty when creating the connector distribution jars.
https://github.com/apache/flink-connector-aws/tree/main/flink-connector-aws/flink-connector-dynamodb/src/test/java/org/apache/flink/connector/dynamodb/sink/examples

We could do something similar and point to them via README, OR we could add a maven profile that only adds the examples module for PR / nightlies

nicusX replied:
I moved the example application under tests, as suggested, merging everything in a single class.
The usage of the AMP connector is commented out, being a separate dependency, but still present to show the usage, in case someone wants to use it.

https://github.com/apache/flink-connector-prometheus/blob/5afb890aa8fe860ce25bcf89f7cdee93a6d2c4be/flink-connector-prometheus/src/test/java/org/apache/flink/connector/prometheus/sink/examples/DataStreamExample.java

hlteoh37 commented on README.md:
```diff
@@ -8,6 +8,14 @@ Apache Flink is an open source stream processing framework with powerful stream-
+
+This repository contains the following modules
+
+* [Prometheus Connector](./prometheus-connector): Flink Prometheus Connector implementation; supports optional request signer
```
nit: can we name this `flink-connector-prometheus` to follow the repo / artifact name?

nicusX replied:
Renamed subdirectories to match artifact names

hlteoh37 commented on README.md:
```diff
@@ -8,6 +8,14 @@ Apache Flink is an open source stream processing framework with powerful stream-
+
+* [Prometheus Connector](./prometheus-connector): Flink Prometheus Connector implementation; supports optional request signer
+* [Sample application](./example-datastream-job): Sample application showing the usage of the connector with DataStream API. It also demonstrates how to configure the request signer. 
+* [Amazon Managed Prometheus Request Signer](./amp-request-signer): Implementation of request signer for Amazon Managed Prometheus (AMP)
```
nit: Should we name this as `request-signer-amp` - this might read a little nicer

nicusX replied:
It was actually requested earlier in the PR to make the module name fully qualified, prepending `flink-connector-prometheus-`.
I am renaming the module to `flink-connector-prometheus-request-signer-amp` to make it fully qualified and retaining the logical order, general to specific

hlteoh37 commented on amp-request-signer/README.md:
```diff
@@ -0,0 +1,31 @@
+### Sample usage
+
+To enable request signing for Amazon Managed Prometheus, and instance of `AmazonManagedPrometheusWriteRequestSigner`
+must
```
nit: why is this on its own line?

nicusX replied:
Fixed

hlteoh37 reviewed (commented):
Added some initial feedback!

nicusX commented:
Addressed the latest comments and ported AMP signer to AWS SDKv2 (!)

nicusX commented:
I moved the example application under tests and removed the separate module completely.
I added links to the example application in the README.

hlteoh37 commented on README.md:
```diff
@@ -2,12 +2,27 @@
+
+* [Prometheus Connector](./flink-connector-prometheus): Flink Prometheus Connector implementation; supports optional
+  request signer
+* [Sample application](./example-datastream-job): Sample application showing the usage of the connector with DataStream
```
This needs to change.

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/README.md:
```diff
@@ -0,0 +1,30 @@
+// Build the sink to AMP using the request signer
+AsyncSinkBase<PrometheusTimeSeries, Types.TimeSeries> sink = PrometheusSink.builder()
+        .setPrometheusRemoteWriteUrl(prometheusRemoteWriteUrl)
+        .setRequestSigner(new AmazonManagedPrometheusWriteRequestSigner(prometheusRemoteWriteUrl, prometheusRegion))
```
LOVE how easy this interface is! :D

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/pom.xml:
```diff
@@ -0,0 +1,70 @@
+            <dependency>
+                <groupId>software.amazon.awssdk</groupId>
+                <artifactId>bom</artifactId>
+                <version>${aws.sdkv2.version}</version>
```
nice! :D

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AmazonManagedPrometheusWriteRequestSigner.java:
```diff
@@ -0,0 +1,86 @@
+    public AmazonManagedPrometheusWriteRequestSigner(String remoteWriteUrl, String awsRegion) {
+        Preconditions.checkArgument(
+                StringUtils.isNotBlank(awsRegion), "Missing or blank AMP workspace region");
+        Preconditions.checkNotNull(
```
I think this is supposed to be `Preconditions.checkArgument(` . We should also have unit tests for this.

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AmazonManagedPrometheusWriteRequestSigner.java:
```diff
@@ -0,0 +1,86 @@
+     */
+    public AmazonManagedPrometheusWriteRequestSigner(String remoteWriteUrl, String awsRegion) {
+        Preconditions.checkArgument(
+                StringUtils.isNotBlank(awsRegion), "Missing or blank AMP workspace region");
```
nit: we can phrase it something like `remoteWriteUrl cannot be null or empty`

hlteoh37 reviewed (commented):
Added some comments on the request signer. Much nicer now it uses SDK v2! Thanks @nicusX

hlteoh37 commented on flink-connector-prometheus/README.md:
```diff
@@ -0,0 +1,260 @@
+Due to the strict [ordering](https://prometheus.io/docs/concepts/remote_write_spec/#ordering)
+and [format](https://prometheus.io/docs/concepts/remote_write_spec/#labels) requirements
+of Prometheus Remote-Write, the sink guarantees that input data are written to Prometheus only if input data are in
+order and well-formed.
```
The word "guarantee" here is a little confusing. Maybe we can say something like the "Prometheus Remote-Write has strict guarantees around ordering and format. This sink requires that users ensure that input data is in order and well-formed before adding it to the sink."

hlteoh37 commented on flink-connector-prometheus/README.md:
```diff
@@ -0,0 +1,260 @@
+#### Input data validation
+
+Prometheus imposes strict constraints to the content sent to remote-write, including label format and ordering, sample
+time ordering ecc.
```
nit: etc.

hlteoh37 commented on flink-connector-prometheus/README.md:
```diff
@@ -0,0 +1,260 @@
+        .setMaxTimeInBufferMS(5000)                 // Max time a time-series is buffered for batching (default: 5000 ms)
+        .setRetryConfiguration(RetryConfiguration.builder()
+                .setInitialRetryDelayMS(30L)            // Initial retry delay (default: 30 ms)
+                .setMaxRetryDelayMS(5000L)              // Maximum retray delay, with exponential backoff (default: 5000 ms)
```
nit: retry delay

hlteoh37 commented on flink-connector-prometheus/README.md:
````diff
@@ -0,0 +1,260 @@
+```java
+PrometheusSink sink = PrometheusSink.builder()
+        .setMaxBatchSizeInSamples(500)              // Batch size (write-request size), in samples (default: 500)
+        .setMaxRecordSizeInSamples(500)             // Max sink input record size, in samples (default: 500), must be <= maxBatchSizeInSamples
````
Might be good to clarify the behavior when this is exceeded. The sink will continuously restart and reject the record, to be used for data validation before sending to sink. Might be even mentioning that this is not expected in the normal use case of the sink. (As compared to maxBatchSize which can be tuned for performance reasons)

nicusX replied:
Good point.
I will explain this in the actual docs, in the other PR

nicusX replied:
I also modified the docs, pushing the change to the other PR #2

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusRemoteWriteHttpRequestBuilder.java:
```diff
@@ -0,0 +1,77 @@
+import java.util.Map;
+
+/** Builds the POST request to the Remote-Write endpoint for a given binary payload. */
+public class PrometheusRemoteWriteHttpRequestBuilder {
```
Let's annotate this as `@Internal`

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/HttpResponseCallback.java:
```diff
@@ -0,0 +1,188 @@
+ * limit has been exceeded. It is responsibility of the callback distinguishing between these
+ * conditions.
+ */
+class HttpResponseCallback implements FutureCallback<SimpleHttpResponse> {
```
Let's annotate this as `@Internal`

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSink.java:
```diff
@@ -0,0 +1,148 @@
+                maxBatchSizeInSamples, // maxBatchSize,
+                maxInFlightRequests,
+                maxBufferedRequests,
+                maxBatchSizeInSamples, // maxBatchSizeInBytes,
```
Can we add a comment to explain that we don't need to check the bytes, so we simply use sample as replacement for bytes

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSink.java:
```diff
@@ -0,0 +1,148 @@
+
+    public int getMaxBatchSizeInSamples() {
+        return maxBatchSizeInSamples;
+    }
```
can this be `private`?

hlteoh37 replied:
Actually can we just delete this?

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkBuilder.java:
```diff
@@ -0,0 +1,264 @@
+
+    public String getMetricGroupName() {
+        return metricGroupName;
+    }
```
Can we remove the `get` methods on the builder?

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkBuilder.java:
```diff
@@ -0,0 +1,264 @@
+
+    /** Not supported. Use setMaxBatchSizeInSamples(int) instead */
+    @Override
+    public PrometheusSinkBuilder setMaxBatchSize(int maxBatchSize) {
+        throw new UnsupportedOperationException("maxBatchSize is not supported by this sink");
+    }
+
+    /** Not supported. Use setMaxBatchSizeInSamples(int) instead */
+    @Override
+    public PrometheusSinkBuilder setMaxBatchSizeInBytes(long maxBatchSizeInBytes) {
+        throw new UnsupportedOperationException(
+                "maxBatchSizeInBytes is not supported by this sink");
+    }
+
+    /** Not supported. Use setMaxRecordSizeInSamples(int) instead */
+    @Override
+    public PrometheusSinkBuilder setMaxRecordSizeInBytes(long maxRecordSizeInBytes) {
+        throw new UnsupportedOperationException(
+                "maxRecordSizeInBytes is not supported by this sink");
+    }
+
+    /** Not supported. Use getMaxBatchSizeInSamples() instead */
+    @Override
+    protected Integer getMaxBatchSize() {
+        throw new UnsupportedOperationException("maxBatchSize is not supported by this sink");
+    }
+
+    /** Not supported. Use getMaxBatchSizeInSamples() instead */
+    @Override
+    protected Long getMaxBatchSizeInBytes() {
+        throw new UnsupportedOperationException(
+                "maxRecordSizeInBytes is not supported by this sink");
+    }
+
+    /** Not supported. Use getMaxRecordSizeInSamples() instead */
+    @Override
+    protected Long getMaxRecordSizeInBytes() {
+        throw new UnsupportedOperationException(
+                "maxRecordSizeInBytes is not supported by this sink");
+    }
```
We don't need to do this, as we don't expect this builder to be extended!

nicusX replied:
But if someone does, because they are creating their custom extended sink and they think to reuse this, they will get to the methods exposed by `AsyncSinkBaseBuilder`, potentially causing a mess.
I think shadowing those methods is safer and harmless.

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusTimeSeriesConverter.java:
```diff
@@ -0,0 +1,61 @@
+ * Converts the sink input {@link PrometheusTimeSeries} into the Protobuf {@link Types.TimeSeries}
+ * that are sent to Prometheus.
+ */
+public class PrometheusTimeSeriesConverter
```
`@PublicEvolving`?

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusTimeSeriesLabelsAndMetricNameKeySelector.java:
```diff
@@ -0,0 +1,44 @@
+ * <p>The partition key is the hash of all labels AND the metricName. The metricName is added as
+ * additional label by the sink, before writing to Prometheus.
+ */
+public class PrometheusTimeSeriesLabelsAndMetricNameKeySelector
```
`@PublicEvolving`?

hlteoh37 replied:
Can we also name this something smaller? Like `PrometheusKeySelector`?

nicusX replied:
Do we pay by character in classname? ;)
This is long but it's meant to express exactly what it does: a key selector extracting a compound key with all labels and metric name from a Time Series.
A better name would be explaining why you should use it, but it's even more wordy: "key selector for sink parallelism > 1 to prevent accidental out of order writes"

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/errorhandling/OnErrorBehavior.java:
```diff
@@ -0,0 +1,27 @@
+ * Defines the behaviour when an error is encountered: discard the offending request and continue,
+ * or fail, throwing an exception.
+ */
+public enum OnErrorBehavior {
```
`@PublicEvolving`. Actually let's put this under a `ConfigOption` class

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/errorhandling/PrometheusSinkWriteException.java:
```diff
@@ -0,0 +1,98 @@
+package org.apache.flink.connector.prometheus.sink.errorhandling;
+
+/** Exception during writing to Prometheus Remote-Write endpoint. */
+public class PrometheusSinkWriteException extends RuntimeException {
```
`@Internal`?

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/errorhandling/SinkWriterErrorHandlingBehaviorConfiguration.java:
```diff
@@ -0,0 +1,105 @@
+ * Configure the error-handling behavior of the writer, for different types of error. Also defines
+ * default behaviors.
+ */
+public class SinkWriterErrorHandlingBehaviorConfiguration implements Serializable {
```
`@PublicEvolving`?

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/http/PrometheusAsyncHttpClientBuilder.java:
```diff
@@ -0,0 +1,93 @@
+ * Builder for async http client that will retry, based on the {@link RemoteWriteRetryStrategy}
+ * specified.
+ */
+public class PrometheusAsyncHttpClientBuilder implements Serializable {
```
`@PublicEvolving`?

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/http/RetryConfiguration.java:
```diff
@@ -0,0 +1,79 @@
+import java.io.Serializable;
+
+/** Defines the retry strategy configuration. */
+public class RetryConfiguration implements Serializable {
```
Can we just make this a set of `ConfigOptions`?

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/metrics/SinkMetrics.java:
```diff
@@ -0,0 +1,87 @@
+import org.apache.flink.metrics.MetricGroup;
+
+/** Wraps all metrics in a single class. */
+public class SinkMetrics {
```
`@Internal`?

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/metrics/SinkMetricsCallback.java:
```diff
@@ -0,0 +1,69 @@
+import static org.apache.flink.connector.prometheus.sink.metrics.SinkMetrics.SinkCounter.NUM_WRITE_REQUESTS_PERMANENTLY_FAILED;
+
+/** Callback updating {@link SinkMetrics} on specific request outcomes. */
+public class SinkMetricsCallback {
```
`@Internal`?

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/errorhandling/PrometheusSinkWriteException.java:
```diff
@@ -0,0 +1,98 @@
+        this.httpResponseBody = httpResponseBody;
+    }
+
+    public String getReason() {
+        return reason;
+    }
+
+    public int getHttpStatusCode() {
+        return httpStatusCode;
+    }
+
+    public String getHttpReasonPhrase() {
+        return httpReasonPhrase;
+    }
+
+    public int getTimeSeriesCount() {
+        return timeSeriesCount;
+    }
+
+    public long getSampleCount() {
+        return sampleCount;
+    }
+
+    public String getHttpResponseBody() {
+        return httpResponseBody;
+    }
```
can we remove these?

hlteoh37 commented on pom.xml:
```diff
@@ -0,0 +1,104 @@
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <flink.version>1.18.1</flink.version>
```
Let's make this `1.19.1`

hlteoh37 commented on pom.xml:
```diff
@@ -0,0 +1,104 @@
+
+    <build>
+        <plugins>
+            <plugin>
```
Could we copy over the other plugins here too ? 

https://github.com/apache/flink-connector-aws/blob/main/pom.xml#L421-L477

hlteoh37 reviewed (changes requested):
Thanks for the super detailed contribution @nicusX! Left some final comments

nicusX commented:
@hlteoh37 I addressed all remaining comments above

nicusX commented:
@hlteoh37 I should have addressed all comments. 
The configuration classes have been moved into a top-level `PrometheusSinkConfiguration` class

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/src/test/java/org/apache/flink/connector/prometheus/sink/aws/DummAwsSessionCredentialProvider.java:
```diff
@@ -0,0 +1,29 @@
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials;
+
+/** Dummy AwsCredentialsProvider returning AwsSessionCredentials. */
+public class DummAwsSessionCredentialProvider implements AwsCredentialsProvider {
```
nit: Dummy

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AWS4SignerForAMP.java:
```diff
@@ -0,0 +1,347 @@
+package org.apache.flink.connector.prometheus.sink.aws;
```
Lets ensure that the license headers are added to all classes

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AWS4SignerForAMP.java:
```diff
@@ -0,0 +1,347 @@
+ * Implements Http/1.1 AWS Signature Version 4a request signing for POST requests to the Amazon
+ * Managed Prometheus service (`aps`).
+ */
+public class AWS4SignerForAMP {
```
Also let's make this `@PublicEvolving`

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AmazonManagedPrometheusWriteRequestSigner.java:
```diff
@@ -0,0 +1,106 @@
+ * <p>On every request, AWS credentials are retrieved using an AwsCredentialsProvider, and used to
+ * sign the request, with AWS Signature Version 4a.
+ */
+public class AmazonManagedPrometheusWriteRequestSigner implements PrometheusRequestSigner {
```
Headers + `@PublicEvolving`

hlteoh37 commented on flink-connector-prometheus/README.md:
```diff
@@ -0,0 +1,259 @@
+* `numWriteRequestsPermanentlyFailed` number of write requests permanently failed, due to any reasons (non retryable,
+  max nr of retries)
+
+**Note**: the `numByteSend` does not measure the number of bytes, due to an internal limitation of the base sink.
```
nit `numBytesOut`?

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/src/test/java/org/apache/flink/connector/prometheus/sink/aws/RequestHeaderTestUtil.java:
```diff
@@ -0,0 +1,21 @@
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class RequestHeaderTestUtil {
```
License header + `@Internal`

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/src/test/java/org/apache/flink/connector/prometheus/sink/aws/AWS4SignerForAMPTest.java:
```diff
@@ -0,0 +1,73 @@
+package org.apache.flink.connector.prometheus.sink.aws;
```
License headers

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/src/test/java/org/apache/flink/connector/prometheus/sink/aws/AmazonManagedPrometheusWriteRequestSignerTest.java:
```diff
@@ -0,0 +1,112 @@
+package org.apache.flink.connector.prometheus.sink.aws;
```
License headers

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/src/test/java/org/apache/flink/connector/prometheus/sink/aws/DummAwsSessionCredentialProvider.java:
```diff
@@ -0,0 +1,29 @@
+package org.apache.flink.connector.prometheus.sink.aws;
```
License headers

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/src/test/java/org/apache/flink/connector/prometheus/sink/aws/DummyAwsBasicCredentialsProvider.java:
```diff
@@ -0,0 +1,25 @@
+package org.apache.flink.connector.prometheus.sink.aws;
```
License headers

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/src/test/java/org/apache/flink/connector/prometheus/sink/aws/RequestHeaderTestUtil.java:
```diff
@@ -0,0 +1,21 @@
+package org.apache.flink.connector.prometheus.sink.aws;
```
License headers

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkBuilder.java:
```diff
@@ -0,0 +1,241 @@
+import java.util.Optional;
+
+/** Builder for Sink implementation. */
+public class PrometheusSinkBuilder
```
`@PublicEvolving`

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkConfiguration.java:
```diff
@@ -0,0 +1,165 @@
+package org.apache.flink.connector.prometheus.sink;
```
License headers.

Thanks for creating this class!

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkConfiguration.java:
```diff
@@ -0,0 +1,165 @@
+     * Configure the error-handling behavior of the writer, for different types of error. Also
+     * defines default behaviors.
+     */
+    public static class SinkWriterErrorHandlingBehaviorConfiguration implements Serializable {
```
We'll have to annotate every class in this file with `@PublicEvolving`

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkException.java:
```diff
@@ -0,0 +1,33 @@
+ * A {@link RuntimeException} wrapper indicating a generic unhandled exception was thrown by the
+ * Prometheus sink.
+ */
+public class PrometheusSinkException extends RuntimeException {
```
`@Internal`?

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkWriter.java:
```diff
@@ -0,0 +1,198 @@
+ * the number of Samples in each TimeSeries to the max batch size, setting maxRecordSizeInBytes =
+ * maxBatchSizeInBytes.
+ */
+public class PrometheusSinkWriter extends AsyncSinkWriter<PrometheusTimeSeries, Types.TimeSeries> {
```
`@Internal`

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusStateSerializer.java:
```diff
@@ -0,0 +1,154 @@
+ * Serializes/deserializes the sink request-entry, the protobuf-generated {@link Types.TimeSeries},
+ * using protobuf.
+ */
+public class PrometheusStateSerializer extends AsyncSinkWriterStateSerializer<Types.TimeSeries> {
```
`@Internal`

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/RequestEntrySizeUtils.java:
```diff
@@ -0,0 +1,63 @@
+import java.util.Collection;
+
+/** Collection of methods to calculate the sink RequestEntry "size". */
+public class RequestEntrySizeUtils {
```
`@Internal`

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/http/RemoteWriteResponseClassifier.java:
```diff
@@ -0,0 +1,38 @@
+import org.apache.hc.core5.http.HttpResponse;
+
+/** Classify http responses based on the status code. */
+public class RemoteWriteResponseClassifier {
```
`@Internal`

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/http/RemoteWriteRetryStrategy.java:
```diff
@@ -0,0 +1,129 @@
+ * {@link UnknownHostException}, {@link ConnectException}, {@link NoRouteToHostException}, and
+ * {@link SSLException}.
+ */
+public class RemoteWriteRetryStrategy implements HttpRequestRetryStrategy {
```
`@Internal`

hlteoh37 reviewed (changes requested):
Changes look good, Thanks @nicusX . Left 1 more round of sanity check to validate that we have license headers and `@` annotations on each class. The annotations are important so that we know in the future which classes are public interfaces and automatically check for backwards compatibility!

nicusX commented:
@hlteoh37 all non-test classes are now annotated with either `@PublicEvolving` or `@Internal`. 
All .java, .xml, and .properties file have the Apache 2.0 license header.
Note that the .proto files, that have been copied into the project just for reference, have their own license header

hlteoh37 reviewed (approved):
LGTM Thanks for the contribution @nicusX ! Can we squash the commits into one please?
